keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,575 @@
+important benefits:
+
+- It's more convenient for types authors.
+- The types and their behaviors are easier to understand since they are
+described together.
+- They are less error-prone, and discourage people from using anti-patterns
+like "cross-field dependency defaults" (default depends on another
+field), non-deterministic or complicated defaults. These patterns
+might be useful in limited circumstances, but it should be easier to do
+the right thing in most cases.
+- Removes the inconsistency with CRDs that already support this
+mechanism.
+- Solves bug where associative keys needs to either be required or
+defaulted. Default keys will now be able to be specified in the OpenAPI
+for built-in types, and SSA will be able to fill in the missing keys.
+
+### Goals
+
+The goal is to add a new `// +default` marker to our current built-in Go
+IDL. That marker will be transformed into the OpenAPI `default` field
+and then routed to defaulting functions so that defaulting can be done
+declaratively. This new marker would also be introduced in kubebuilder,
+and we will make the behavior equivalent between built-in types and CRDs
+so that built-in types can be directly used in kubebuilder, and the
+declarative defaults will be automatically carried along.
+
+### Non-Goals
+
+While this proposal is probably paving the way for additional OpenAPI
+based validation, validation markers for built-in is entirely
+out-of-scope for this project.
+
+This proposal also doesn't intend to remove the existing defaulting
+mechanism since some defaulting still requires non-static values that
+need to be computed and can't be defined declaratively. We could replace
+some defaulting that doesn't require to be done as code though.
+
+## Proposal
+
+It is important to us that the OpenAPI definition of a type should
+result in the same behavior, whether the type is built-in or a CRD.
+
+In order to do that, we need to change how CRDs are defaulted, and
+create a defaulting mechanism for built-in that would behave the same
+way.
+
+### CRD Normalization
+
+Null values in CRDs are currently not defaulted. On the other hand,
+`null` typically means "Devoid of
+value"(link)[https://yaml.org/type/null.html] in json/yaml, though it is
+treated as different from not present ("a null value is valid and
+different from not having that key in the mapping"). JSon Merge patch
+interprets a "null" value for a field as a means to delete it: "Null
+values in the merge patch are given special meaning to indicate the
+removal of existing values in the
+target."(link)[https://tools.ietf.org/rfc/rfc7386.txt#:~:text=Null%20values%20in%20the%20merge]
+
+On the other hand, while `{field: null}` and `{}` probably have the same
+meaning, they do not "look" the same and are hard to compare for
+clients. They are also currently treated differently by Kubernetes for
+defaulting purposes.
+
+We propose the creation of a new CRD flag: `preserveNullValues` which
```

Comment:
nit, `preserveNullValues: true` here and `x-kubernetes-remove-nulls: false` later... which one?

Reply:
I assume use of this is mutually exclusive with use of nullable anywhere in the object?

Reply:
Yeah, that's a mistake. And as soon as I started looking at the actual implementation, I noticed that "PreserveUnknownFields" is deprecated in the CRD spec. I'll update but it should match the preserve-unknown-fields which is recommended in the openapi (per version).